Public/Src/Cache/ContentStore/Vsts/DedupReadOnlyContentSession.cs

```diff
@@ -155,14 +172,62 @@ public class DedupReadOnlyContentSession : ContentSessionBase
                 return new PinResult($"DedupStore client requires {RequiredHashType}. Cannot take HashType '{contentHash.HashType}'.");
             }
 
-            try
+            var pinResult = CheckPinInMemory(contentHash);
+            if (pinResult.Succeeded)
             {
-                var pinResult = CheckPinInMemory(contentHash);
-                if (pinResult.Succeeded)
-                {
-                    return pinResult;
-                }
+                return pinResult;
+            }
+
+            // Since pinning the whole tree can be an expensive operation, we have optimized how we call it. Depending on the current
+            //  keepUntil of the root node, which is unexpensive to check, the operation will behave differently:
+            //      The pin operation will be ignored if it is greater than ignorePinThreshold, to reduce amount of calls
+            //      The pin operation will be inlined if it is lower than pinInlineThreshold, to make sure that we don't try to use
+            //          content that we pin in the background but has expired before we could complete the pin.
+            //      The pin operation will be done asynchronously and will return success otherwise. Most calls should follow this
+            //          behavior, to avoid waiting on a potentially long operation.
+
+            VstsBlobIdentifier blobId = ToVstsBlobIdentifier(contentHash.ToBlobIdentifier());
+            VstsDedupIdentifier dedupId = blobId.ToDedupIdentifier();
+            var keepUntilResult = await CheckNodeKeepUntilAsync(context, dedupId);
+            if (!keepUntilResult.Succeeded)
+            {
+                // Returned a service error. Fail fast.
+                return new PinResult(keepUntilResult);
+            }
+            else if (!keepUntilResult.Value.HasValue)
+            {
+                // Content not found.
+                return new PinResult(PinResult.ResultCode.ContentNotFound);
+            }
 
+            var timeLeft = keepUntilResult.Value.Value - DateTime.UtcNow;
+
+            if (timeLeft > _ignorePinThreshold)
+            {
+                Tracer.Debug(context, $"Pin ignored bacause keepUntil has remaining time [{timeLeft}] that is greater than the ignorePinThreshold");

```

Comment:
Maybe it make sense to call it "Pin was skipped", instead of "ignored"? I'm fine with the current terms, but I'm curious maybe "skip" here is a better term.